File: src/main/java/net/jeikobu/jbase/AbstractBot.java
package net.jeikobu.jbase;

import net.jeikobu.jbase.command.AbstractCommand;
import net.jeikobu.jbase.command.CommandManager;
import net.jeikobu.jbase.config.AbstractConfigManager;
import net.jeikobu.jbase.impl.commands.ChangeLocaleCommand;
import net.jeikobu.jbase.impl.commands.ChangePrefixCommand;
import sx.blah.discord.api.ClientBuilder;
import sx.blah.discord.api.IDiscordClient;
import sx.blah.discord.api.events.EventDispatcher;

import java.util.Arrays;
import java.util.List;

@SuppressWarnings("WeakerAccess")
public abstract class AbstractBot {
    private final List<Class<? extends AbstractCommand>> defaultCommands = Arrays
            .asList(ChangeLocaleCommand.class, ChangePrefixCommand.class);

    private final IDiscordClient client;
    private final EventDispatcher dispatcher;
    private final CommandManager commandManager;
    private final AbstractConfigManager configManager;

    public AbstractBot(ClientBuilder clientBuilder, AbstractConfigManager configManager) {
        commandManager = new CommandManager(configManager);
        this.configManager = configManager;

        if (configManager.getGlobalConfig().useDefaultCommands()) {
            registerDefaultCommands();
        }

        clientBuilder.withToken(configManager.getGlobalConfig().getToken());

        client = clientBuilder.build();
        dispatcher = client.getDispatcher();
        dispatcher.registerListener(commandManager);
    }

    public void registerDefaultCommands() {
        commandManager.registerAll(defaultCommands);
    }

    public void deregisterDefaultCommands() {
        commandManager.deregisterAll(defaultCommands);
    }

    public IDiscordClient getClient() {
        return client;
    }

    public EventDispatcher getDispatcher() {
        return dispatcher;
    }

    public CommandManager getCommandManager() {
        return commandManager;
    }

    public AbstractConfigManager getConfigManager() {
        return configManager;
    }
}

File: src/main/java/net/jeikobu/jbase/AbstractBot.kt
package net.jeikobu.jbase

import net.dv8tion.jda.core.JDA
import net.dv8tion.jda.core.JDABuilder
import net.jeikobu.jbase.command.AbstractCommand
import net.jeikobu.jbase.command.CommandManager
import net.jeikobu.jbase.config.AbstractConfigManager
import net.jeikobu.jbase.impl.commands.ChangeLocaleCommand
import net.jeikobu.jbase.impl.commands.ChangePrefixCommand
import java.util.*
import kotlin.reflect.KClass

abstract class AbstractBot(configManager: AbstractConfigManager) {
    protected val defaultCommands: MutableList<KClass<out AbstractCommand>> = Arrays.asList(ChangeLocaleCommand::class, ChangePrefixCommand::class)
    protected val commandManager = CommandManager(configManager)
    private val clientBuilder = JDABuilder()

    val client: JDA

    init {
        if (configManager.globalConfig.useDefaultCommands) {
            registerDefaultCommands()
        }

        clientBuilder.setToken(configManager.globalConfig.token)
        clientBuilder.addEventListener(commandManager)

        client = clientBuilder.build()
        client.awaitReady()
    }

    fun registerDefaultCommands() {
        commandManager.registerAll(defaultCommands)
    }

    fun deregisterDefaultCommands() {
        commandManager.deregisterAll(defaultCommands)
    }
}